# Learning and practicing NumPy with examples and exercises

# Mi Cuaderno Interactivo: Explorando NumPy y Programación en Python

¡Bienvenidos! En este cuaderno, he reunido los puntos clave del taller que realicé, donde abordo:

- **Operaciones fundamentales con NumPy**: Para manipular y examinar datos de forma eficiente.

- **Ejemplo de descenso de gradiente en regresión lineal**: Para entender cómo se optimizan parámetros en problemas de aprendizaje automático.

Mi objetivo es mostrar, paso a paso, cómo aprovechar al máximo NumPy y sentar bases sólidas en programación y análisis de datos.

# **1 Ejemplos: NumPy**

## 1. Indexación Básica

Este ejemplo muestra cómo acceder a un elemento específico de un array unidimensional utilizando su índice.

In [35]:
import numpy as np

arr= np.array([10,20,30,40,50])
print(arr[2])


30


`arr[2]` accede al tercer elemento del array (**índice 2**), que es `30`.
## 2.Slicing
Este ejemplo demuestra cómo extraer una subsección de un array unidimensional utilizando slicing.

In [36]:
arrr=np.array([10,20,30,40,50])
print(arr[1:4])

[20 30 40]


`arr[1:4]` selecciona los elementos desde el **índice 1** hasta el **índice 3** (el **índice 4** no se incluye).

## 3.Indexación en Arrays 2D
Este ejemplo muestra cómo acceder a un elemento específico en un array bidimensional utilizando índices de fila y columna

In [37]:
arr= np.array([[1,2,3],[4,5,6],[7,8,9]])
print(arr[1,2])

6


`arr[1, 2]` accede al elemento en la segunda fila (**índice 1**) y tercera columna (**índice 2**), que es `6`.
## 4.Slicing en Arrays 2D
Este ejemplo muestra cómo extraer una submatriz de un array bidimensional utilizando slicing.

In [38]:
arr=np.array([[1,2,3],[4,5,6,],[7,8,9,]])
print(arr[:2,1:])

[[2 3]
 [5 6]]


`arr[:2, 1:]` selecciona una submatriz utilizando slicing:
- `:2` (para las filas): Selecciona las filas desde el inicio (**índice 0**) hasta la fila con **índice 1** (el **índice 2** no se incluye). Esto selecciona las dos primeras filas.
- `1:` (para las columnas): Selecciona las columnas desde el **índice 1** hasta el final. Esto selecciona las columnas 2 y 3.
 #### **resultado:** 

 $$
\begin{bmatrix}
2 & 3  \\
5 & 6 
\end{bmatrix}
$$

## 5.Indexación en Tensores 3D
Este ejemplo muestra cómo acceder a elementos específicos y submatrices en un tensor tridimensional.

In [39]:
tensor=np.array([[[1,2,3],[4,5,6,]],[[7,8,9],[13,11,12]],[[13,14,15],[16,17,18]]])
print(tensor[1,0,2])

9


- `tensor[1, 0, 2]` accede al elemento en la segunda matriz (**índice 1**), primera fila (**índice 0**), y tercera columna (**índice 2**), que es `9`.



In [40]:
print(tensor[1])

[[ 7  8  9]
 [13 11 12]]


- `tensor[1]` selecciona toda la segunda matriz (**índice 1**), que es 
$$
\begin{bmatrix}
7 & 8 & 9 \\
13 & 11 & 12
\end{bmatrix}
$$
## 6.Slicing en Tensores 3D
Este ejemplo muestra cómo extraer submatrices de un tensor tridimensional utilizando slicing.


In [41]:
tensor=np.arange(27).reshape(3,3,3)
print(tensor[1])




[[ 9 10 11]
 [12 13 14]
 [15 16 17]]


- `np.arange(27)` genera una lista de 27 numeros
- `.reshape(3,3,3)` convierte esa lista en un tensor, lo que tenemos son 3 matrices, cada una con 3 filas y 3 columnas.
- `tensor[1]` selecciona toda la segunda matriz (**índice 1**).


In [42]:
print(tensor[:,1,:])

[[ 3  4  5]
 [12 13 14]
 [21 22 23]]


- `tensor[:, 1, :]` selecciona la segunda fila (**índice 1**) de todas las matrices, formando una submatriz 3x3, que es 
\begin{bmatrix}
3& 4 & 5 \\
12 & 13 & 14\\
21 & 22 & 23
\end{bmatrix}
$$

In [43]:
print(tensor[:,:,2])

[[ 2  5  8]
 [11 14 17]
 [20 23 26]]


- `tensor[:, :, 1]` selecciona la tercera columna (**índice 2**) de todas las matrices, formando una submatriz 3x3, que es 
$$
\begin{bmatrix}
2 & 5 & 8 \\
11 & 14 & 17\\
20 & 23 & 26
\end{bmatrix}
$$

## 7. 4D Tensor Indexing
En este ejemplo se crea un tensor de 4 dimensiones y se accede a elementos y bloques específicos dentro de él.

In [44]:
tensor=np.arange(48).reshape(2,3,2,4)
print(tensor[1,2,0,3])


43


- `np.arange(48)` genera una lista de numeros del 0 al 47
- `reshape(2,3,2,4)` le da forma de tensor de 4 dimensiones: 2 bloques, 3 filas, 2 columnas y 4 valores por fila
- `tensor[1, 2, 0, 3]` accede al valor ubicado en el segundo bloque, tercera fila, primera subfila, cuarta posición → da como resultado `43`.

In [45]:
print(tensor[1,2])

[[40 41 42 43]
 [44 45 46 47]]


- `tensor[1, 2]` accede a todo el bloque en la posición [1, 2], devolviendo una matriz de **2x4**.

\begin{bmatrix}
40 & 41 & 42 & 43 \\
44 & 45 & 46 & 47
\end{bmatrix}
$$

## 8.Slicing a 4D Tensor
En este ejemplo se realiza slicing sobre un tensor 4D para obtener subconjuntos a lo largo de diferentes dimensiones.

In [46]:
tensor=np.arange(48).reshape(2,3,2,4)
print(tensor[1])


[[[24 25 26 27]
  [28 29 30 31]]

 [[32 33 34 35]
  [36 37 38 39]]

 [[40 41 42 43]
  [44 45 46 47]]]


- `tensor[1]` selecciona el segundo bloque completo (de forma 3x2x4).
$$
\left[
\begin{array}{c}
\begin{bmatrix}
24 & 25 & 26 & 27 \\
28 & 29 & 30 & 31
\end{bmatrix} \\
\\
\begin{bmatrix}
32 & 33 & 34 & 35 \\
36 & 37 & 38 & 39
\end{bmatrix} \\
\\
\begin{bmatrix}
40 & 41 & 42 & 43 \\
44 & 45 & 46 & 47
\end{bmatrix}
\end{array}
\right]
$$





In [47]:
print(tensor[:,2,:,:])

[[[16 17 18 19]
  [20 21 22 23]]

 [[40 41 42 43]
  [44 45 46 47]]]


- `tensor[:, :, 0, :]` selecciona la primera subfila (dimensión 3) de cada fila de cada bloque → forma final **2x3x4**.
$$
\left[
\begin{array}{c}
\begin{bmatrix}
16 & 17 & 18 & 19 \\
20 & 21 & 22 & 23
\end{bmatrix} \\
\\
\begin{bmatrix}
40 & 41 & 42 & 43 \\
44 & 45 & 46 & 47
\end{bmatrix}
\end{array}
\right]
$$

## 9. Matrix Addition
En este ejemplo se suman dos matrices del mismo tamaño elemento por elemento.

In [48]:
A=np.array([[1,2],[3,4]])
B=np.array([[5,6],[7,8]])
print(A+B)

[[ 6  8]
 [10 12]]


- Se definen dos matrices 2x2: **A** y **B**.

- La suma A + B suma cada elemento correspondiente:

`[[1+5, 2+6], [3+7, 4+8]] → [[6, 8], [10, 12]]`

- El resultado es una matriz del mismo tamaño con los valores sumados.

## 10. Matrix Multiplication
En este ejemplo se realiza la multiplicación matricial (producto punto) entre dos matrices compatibles.

In [49]:
A=np.array([[1,2],[3,4]])
B=np.array([[5,6],[7,8]])
result=np.dot(A,B)
print(result)

[[19 22]
 [43 50]]


- `np.dot(A, B)` aplica el producto punto entre matrices:

La forma de ambas es 2x2, por lo que son compatibles.

Resultado:
\begin{bmatrix}
19 & 22  \\
43 & 50
\end{bmatrix}
$$

## 11. Transpose of a Matrix
En este ejemplo se obtiene la transpuesta de una matriz, intercambiando filas por columnas.

In [50]:
A=np.array([[1,2],[3,4]])
result=np.transpose(A)
print(result)

[[1 3]
 [2 4]]


- `np.transpose(A)` invierte las filas y columnas de la matriz **A**.

Original:
$$
\begin{bmatrix}
1 & 2 \\
3 & 4 
\end{bmatrix}
$$

Transpuesta:
$$
\begin{bmatrix}
1 & 3 \\
2 & 4 
\end{bmatrix}
$$


## 12. Matrix Determinant and Inverse
En este ejemplo se calcula el determinante y la inversa de una matriz cuadrada.

In [51]:
A=np.array([[1,2],[3,4]])
det=np.linalg.det(A)
inverse=np.linalg.inv(A)
print('Determinant:',det)
print('Inverse: ',inverse)

Determinant: -2.0000000000000004
Inverse:  [[-2.   1. ]
 [ 1.5 -0.5]]


- `np.linalg.det(A)` calcula el determinante de **A**:
`1×4 - 2×3 = -2`

- `np.linalg.inv(A)` calcula la matriz inversa

Resultado:
$$
\begin{bmatrix}
-2 & 1 \\
1.5 & -0.5 
\end{bmatrix}
$$

## 13. Eigenvalues and Eigenvectors
En este ejemplo se calculan los valores propios **(**eigenvalues**) y vectores propios (**eigenvectors**) de una matriz cuadrada.

In [52]:
A=np.array([[1,-1],[1,3]])
aigenvalues, eigenvectors = np.linalg.eig(A)
print('Eigenvalues:', aigenvalues)
print('Eigenvectors:', eigenvectors)

Eigenvalues: [2.00000002 1.99999998]
Eigenvectors: [[-0.70710677 -0.70710679]
 [ 0.70710679  0.70710677]]


- `np.linalg.eig(A)`devuelve:

- `eigenvalues`: una lista con los valores propios λ.

- `eigenvectors`: una matriz donde cada columna es el vector propio correspondiente a cada 
𝜆.

## 14. Solving Linear Equations
En este ejemplo se resuelve un sistema de ecuaciones lineales de la forma $$ Ax= b $$.

In [53]:
A=np.array([[2,1],[1,1]])
b=np.array([3,2])
solution=np.linalg.solve(A,b)
print('Solution:',solution)

Solution: [1. 1.]


- Se tiene un sistema de ecuaciones lineales:

$$
\begin{cases}
2x + y = 3 \\
x + y = 2
\end{cases}
$$

- `np.linalg.solve(A, b)` resuelve este sistema usando álgebra lineal.

- El resultado es el valor de \( x \) que satisface ambas ecuaciones:

$$
x = \begin{bmatrix} 1 \\ 1 \end{bmatrix}
$$


## 15. Gradient Descent Optimization
En este ejemplo se aplica el algoritmo de descenso del gradiente para ajustar parámetros y minimizar el error cuadrático medio (MSE) de un modelo de regresión lineal.

In [54]:
np.random.seed(0)
x=2*np.random.rand(100,1)
y=4+3*x+np.random.randn(100,1)

- Se generan 100 datos aleatorios para x entre 0 y 2.

- Se genera `y` como una línea recta: $$y=4+3x+ruido$$


- El ruido se agrega para simular variabilidad real.

In [55]:
x_b=np.c_[np.ones((100,1)),x]


- Se agrega una columna de unos a `x` para incluir el término independiente (intercepto) en el modelo.

- Resultado: una matriz de forma (100, 2) donde la primera columna es 1 y la segunda es `x`.

In [56]:
learning_rate=0.1
n_iterations=1000

- `learning_rate`: tasa de aprendizaje.

- `n_iterations`: número de iteraciones del ciclo de entrenamiento.

In [57]:
theta = np.random.randn(2, 1)

- Se inicializa aleatoriamente el vector de parámetros 
**theta** con forma (2, 1).

In [58]:
for iteration in range(n_iterations):
 gradients = 2 / 100 * x_b.T.dot(x_b.dot(theta)- y)
 theta = theta- learning_rate * gradients
print("Final theta:", theta)

Final theta: [[4.22215108]
 [2.96846751]]



- Calcula el gradiente del error cuadrático medio.

- Actualiza los parámetros theta moviéndose en la dirección opuesta al gradiente.

- Se repite 1000 veces.
- El modelo aprendió una línea recta cercana a:
$$y=4.22+2.97x$$

# **2.Examples:python functions and classes**

## 1.Simple Function
 Esta función genera un saludo personalizado al recibir un nombre como argumento.

In [59]:
def greet(name):
    return "Hello, " + name + "!"

print(greet("Ana"))  # Output: Hello, Ana!

Hello, Ana!


- La función `greet` toma un parámetro `name`.

- Retorna una cadena que concatena `"Hello, "` con el nombre y un signo de exclamación.

- Al ejecutar` greet("Ana")`, el resultado es `"Hello, Ana!"`.

## 2. Function with Parameters and Return
 Esta función suma dos números y devuelve el resultado.

In [60]:
def add_numbers(a,b):
    sum= a+b
    return sum
result=add_numbers(5,7)
print(result)

12


- `add_numbers` recibe dos parámetros `a` y `b`.

- Suma ambos valores y almacena el resultado en la variable `sum`.

- Devuelve el valor de `sum`.

- En este caso, `5 + 7 = 12`, por lo que el resultado es `12`.



## 3. Function with Default Parameter
 Esta función eleva un número (`base`) a una potencia (`exponent`). Si no se proporciona el exponente, se utiliza 2 por defecto.



In [61]:
def power(base, exponent=2):
    return base ** exponent
print(power(3))  # Output: 9 (3^2)
print(power(2,3))

9
8


- Si solo se pasa un valor (`3`), se eleva al cuadrado por defecto: `3 ** 2 = 9`.

- Si se pasan ambos valores (`2` y `3`), se calcula `2 ** 3 = 8`.

## 4. Function with Multiple Returns
 Esta función recibe una lista de números y devuelve el mínimo y el máximo.


In [62]:
def min_max(numbers):
    
    return min(numbers),max(numbers)
nums=[4,9,2,7,5]
minimum, maximum = min_max(nums)
print(minimum, maximum)  # Output: (2, 9)

2 9


- La función usa `min()` y `max()` para encontrar el valor más bajo y más alto de la lista.

- Devuelve ambos valores como una tupla.

- En este caso, el mínimo es `2` y el máximo es `9`.

##  5. Function with Docstring
 Esta función calcula el área de un rectángulo usando su largo y ancho, e incluye una descripción detallada en el docstring.

In [63]:
def area_of_rectangle(length, width):
 
 """
 Calculates the area of a rectangle.
 Args:
 length (float): The length of the rectangle.
 width (float): The width of the rectangle.
 Returns:
 float: The area of the rectangle.
 """
 return length * width

- El docstring explica qué hace la función, qué argumentos recibe y qué retorna.

- Multiplica `length` por `width` para calcular el área.

## 6. Creación de una clase simple: `Dog`
Esta clase representa un perro con nombre y edad, y tiene un método para ladrar.

In [64]:
class Dog:
 def __init__(self, name, age):

  
    self.name = name
    self.age = age
 def bark(self):
   
    print(f"{self.name} is barking!")
 # Create an instance of the Dog class
dog1 = Dog("Buddy", 3)
dog1.bark() # Output: Buddy is barking!

Buddy is barking!


- `__init__` es el constructor que inicializa los atributos `name` y `age`.

- El método `bark `imprime un mensaje indicando que el perro está ladrando.

- Al crear un perro llamado `"Buddy"` de 3 años y llamar `bark()`, se imprime` "Buddy is barking!"`.

## 7. Agregar métodos y atributos: Student
 Esta clase representa a un estudiante que puede tener una lista de calificaciones y calcular el promedio.

In [65]:
class Student:
    def __init__(self, name, age):
        self.name = name
        self.age = age
        self.grades = []

    def add_grade(self, grade):
        self.grades.append(grade)

    def get_average_grade(self):
        return sum(self.grades) / len(self.grades)

# Crear una instancia de Student y agregar calificaciones
student1 = Student("Alice", 18)
student1.add_grade(85)
student1.add_grade(92)

average_grade = student1.get_average_grade()
print(f"{student1.name}'s average grade: {average_grade}")

Alice's average grade: 88.5


- La clase tiene atributos `name`, `age` y `grades`.

- `add_grade` agrega una calificación a la lista.

- `get_average_grade` calcula el promedio de calificaciones.

- El estudiante `"Alice"` obtiene un promedio de `(85 + 92) / 2 = 88.5`.

## 8. Herencia en clases
Esta sección demuestra cómo una clase puede heredar de otra para compartir y extender funcionalidades.

In [66]:
class Animal:
    def __init__(self, name):
        self.name = name

    def speak(self):
        pass  # Placeholder for subclasses to override

class Cat(Animal):
    def speak(self):
        print(f"{self.name} says Meow!")

class Dog(Animal):
    def speak(self):
        print(f"{self.name} says Woof!")

# Crear instancias de Cat y Dog
cat = Cat("Whiskers")
dog = Dog("Buddy")

cat.speak()  # Output: Whiskers says Meow!
dog.speak()  # Output: Buddy says Woof!


Whiskers says Meow!
Buddy says Woof!


- `Animal` es la clase base con un método `speak` vacío.

- `Cat` y `Dog` son subclases que sobrescriben el método `speak`.

- Cada clase imprime un sonido característico según el tipo de animal.

# **3 Exercises**
##  3.1 IndexingArraysinNumPy
1. Given theNumPy arrayarr=np.array([3,7,1,9,5]), perform
the following indexing operations:


**a)** Retrieve the third element of thearray


 Creamos un array con los elementos dados.

In [67]:
arr= np.array([3,7,1,9,5])
print(arr[2])

1


El tercer elemento se encuentra en el índice 2 (porque los índices comienzan en 0).

**b)** Slice th earray to get elements from index 1 to index 3.

In [68]:
print(arr[1:3])

[7 1]


El slicing `arr[inicio:fin]` incluye el índice `inicio` pero no incluye el índice `fin`.
$$[7,1]$$

In [69]:
arr[3]=12
print(arr)

[ 3  7  1 12  5]


`arr[3]` corresponde al cuarto elemento (`9`), que será reemplazado por `12`.

## 2. Create a 2D NumPy array `matrix` with the following values

Este ejercicio utiliza una **matriz 2D en NumPy** (una lista de listas) para  acceder a elementos específicos y  hacer **slicing** (submatrices).  
La matriz dada es:
$$
\begin{bmatrix}
1 & 2 & 3 \\
4 & 5& 6 \\
7&8&9
\end{bmatrix}
$$

**a)** Retrieve the element in the second row and third column


In [70]:
arr=np.array([[1,2,3],[4,5,6],[7,8,9]])
print(arr[1,2])

6


En Python, los índices comienzan desde 0.


``arr[1, 2]`` accede a la segunda fila (``[4, 5, 6]``) y a su tercer elemento (``6``).

**b)** Slice the matrix to get the submatrix consisting of the first two rows and all columns

In [71]:
print(arr[0:2,:])

[[1 2 3]
 [4 5 6]]


``arr[0:2, :]`` selecciona desde la fila 0 hasta la 1 (la fila 2 no se incluye).

El ``:`` en las columnas significa todas las columnas.
## 3.2 Indexing and Slicing Higher-Dimensional Tensors
 Consider the following 3D tensor representing a stack of matrices:

$$
\text{tensor} =
\left[
\begin{array}{cccc}
\left[
\begin{array}{cccc}
1 & 2 & 3 & 4 \\
5 & 6 & 7 & 8 \\
9 & 10 & 11 & 12
\end{array}
\right], &
\left[
\begin{array}{cccc}
13 & 14 & 15 & 16 \\
17 & 18 & 19 & 20 \\
21 & 22 & 23 & 24
\end{array}
\right], &
\left[
\begin{array}{cccc}
25 & 26 & 27 & 28 \\
29 & 30 & 31 & 32 \\
33 & 34 & 35 & 36
\end{array}
\right]
\end{array}
\right]
$$




In [72]:
tensor= np.array([[[1,2,3,4],[5,6,7,8],[9,10,11,12]],[[13,14,15,16],[17,18,19,20],[21,22,23,24]],[[25,26,27,28],[29,30,31,32],[33,34,35,36]]])
print(tensor)

[[[ 1  2  3  4]
  [ 5  6  7  8]
  [ 9 10 11 12]]

 [[13 14 15 16]
  [17 18 19 20]
  [21 22 23 24]]

 [[25 26 27 28]
  [29 30 31 32]
  [33 34 35 36]]]


**1.** Retrieve the element at the position (1, 0, 2) within the tensor

In [73]:
print(tensor[1,0,2])

15


- Primer índice ``1`` → segunda "matriz"

- Segundo índice ``0`` → primera fila

- Tercer índice ``2`` → tercer elemento → 15

**2.** Slice the tensor to obtain the 3×4 sub-tensor that corresponds to the second “matrix” (slice along the first dimension)

In [74]:
print(tensor[1])

[[13 14 15 16]
 [17 18 19 20]
 [21 22 23 24]]


- ``tensor[1]`` selecciona la segunda matriz completa (índice 1)

- Tiene forma 3×4: 3 filas, 4 columnas




**3.** Slice the tensor to obtain the 2×4 sub-tensor that corresponds to the last row of the last two matrices

In [75]:
print(tensor[1:,2,:])

[[21 22 23 24]
 [33 34 35 36]]


``1:`` → selecciona las últimas dos matrices

``2`` → tercera fila (índice 2) en cada una de esas matrices

``:`` → todas las columnas




**4.** Retrieve the element at the position (2, 1, 3) within the tensor

In [76]:
print(tensor[2,1,3])

32


``2`` → tercera matriz

``1`` → segunda fila

``3`` → cuarto elemento → ``32``


**5.** Slice the tensor to obtain the 2×4 sub-tensor that corresponds to the first row of the last two matrices

In [77]:
print(tensor[1:,0,:])

[[13 14 15 16]
 [25 26 27 28]]


``1:`` → últimas dos matrices

``0`` → primera fila de cada una

``:`` → todas las columnas

# **3.3 Defining Functions in Python**
**1.** Write a Python function calculate
 average that takes a list of numbers
 as input and returns the average of those numbers.

In [80]:
def calculate_average(numbers):
    a=sum(numbers)/len(numbers)
    return a
data=[4,8,10,6]
resultado=calculate_average(data)
print(resultado) 


7.0


``sum(numbers)`` suma todos los elementos: 4 + 8 + 10 + 6 = 28

``len(numbers)`` cuenta cuántos elementos hay: 4

El promedio es 28 / 4 = ``7.0``

In [83]:
def find_power(base,exponent=2):
    return base**exponent
print(find_power(3))
print( find_power(2, 3))

9
8


``find_power(3)`` → 3^2 = 9, porque usa el valor por defecto ``exponent = 2``

``find_power(2, 3)`` → 2^3 = 8, porque pasamos el exponente manualmente